src/mpi/init/mpir_init.c

```diff
@@ -310,6 +312,8 @@ int MPII_Finalize(MPIR_Session * session_ptr)
     mpi_errno = MPID_Finalize();
     MPIR_ERR_CHECK(mpi_errno);
 
+    qmpi_teardown();
```

Comment:
Always uneasy when we see a finishing routine without the matching `qmpi_setup`.

Reply:
It's just the naming that's different because this one gets called inside `MPI_Finalize`. The init function is called `MPII_qmpi_setup`.

This probably needs something smarter for Sessions where you might create a session, then destroy a session, then create it again.